internal/engine/local/controller.go

```diff
@@ -131,29 +136,56 @@ func (c *Controller) start(ctx context.Context, spec ServeSpec, st store.RStore)
 	}
 	ctx = logger.CtxWithLogHandler(ctx, w)
 
-	statusCh := make(chan Status)
-	sequenceNum := proc.sequenceNum
-	go func() {
-		for status := range statusCh {
-			st.Dispatch(LocalServeStatusAction{
-				ManifestName: spec.ManifestName,
-				SequenceNum:  sequenceNum,
-				Status:       status,
-			})
-		}
-	}()
+	statusCh := make(chan status)
+
+	go processStatuses(statusCh, proc.stoppedCh, st, spec.ManifestName, proc.sequenceNum)
 
 	proc.doneCh = c.execer.Start(ctx, spec.ServeCmd, logger.Get(ctx).Writer(logger.InfoLvl), statusCh)
 }
 
+func processStatuses(
+	statusCh chan status,
+	stoppedCh chan struct{},
+	st store.RStore,
+	manifestName model.ManifestName,
+	sequenceNum int) {
+	stopped := false
+	for {
+		select {
+		case status, ok := <-statusCh:
+			if !ok {
+				return
+			}
+			if stopped {
+				continue
+			}
+			runtimeStatus := status.ToRuntime()
+			if runtimeStatus != "" {
+				// TODO(matt) when we get an error, the dot is red in the web ui, but green in the TUI
+				st.Dispatch(LocalServeStatusAction{
+					ManifestName: manifestName,
+					SequenceNum:  sequenceNum,
+					Status:       runtimeStatus,
+				})
+			}
+		case <-stoppedCh:
```

Comment:
Once it's closed, you[ll want to nil it out, or else you'll be at 100% CPU going through this loop?

Reply:
yeah, though this should be mitigated by it always happening in close proximity to the closing of statusCh

anyway, I've got another branch ready to get rid of stoppedCh and just use sequenceNum